# Machine Learning for Finance Freestyle

In this lab you'll be given the opportunity to apply everything you have learned to build a trading strategy for SP500 stocks. First, let's introduce the dataset you'll be using.

## The Data

Use BigQuery's magic function to pull data as follows:

    Dataset Name: ml4f
    Table Name: percent_change_sp500

The following query will pull 10 rows of data from the table:

In [1]:
%%bigquery df
SELECT 
    *
FROM
    `cloud-training-prod-bucket.ml4f.percent_change_sp500`
LIMIT
    10

In [2]:
df.head()

symbol        Date   Open  Close  tomorrow_close  tomo_close_m_close  \
0   BEAM  2008-03-12  65.27  65.31           64.88               -0.43   
1    ABT  1993-08-13  23.62  24.00           24.87                0.87   
2    BAX  2012-10-01  60.42  60.09           60.47                0.38   
3    BIG  2005-08-15  12.65  12.79           11.97               -0.82   
4    BBY  1990-10-10   5.25   5.38            5.50                0.12   

   close_MIN_prior_5_days  close_MIN_prior_20_days  close_MIN_prior_260_days  \
0                0.972439                 0.972439                  0.972439   
1                0.958333                 0.958333                  0.958333   
2                0.995673                 0.966550                  0.792977   
3                0.964816                 0.908522                  0.788116   
4                0.975836                 0.975836                  0.975836   

   close_MAX_prior_5_days  ...  close_STDDEV_prior_20_days  \
0                1.000766  ...                    0.018391   
1                1.025833  ...                    0.033533   
2                1.017640  ...                    0.015421   
3                0.994527  ...                    0.029383   
4                1.092937  ...                    0.152138   

   close_STDDEV_prior_260_days  \
0                     0.090670   
1                     0.095144   
2                     0.058748   
3                     0.062441   
4                     0.357552   

                              close_values_prior_260  days_on_market  \
0  [80.51, 79.67, 79.28, 80.32, 79.64, 79.87, 80....            7866   
1  [30.62, 30.62, 30.62, 30.5, 31.12, 30.62, 30.3...            2620   
2  [56.43, 54.56, 52.91, 53.47, 54.33, 55.87, 55....            7783   
3  [11.67, 11.41, 11.6, 11.88, 11.94, 11.73, 11.4...            4476   
4  [5.75, 5.88, 6.88, 7.88, 8.25, 8.13, 8.0, 7.88...            1386   

   scaled_change  s_p_scaled_change  normalized_change              company  \
0      -0.006584           0.005127          -0.011711             Beam Inc   
1       0.036250           0.004976           0.031274  Abbott Laboratories   
2       0.006324           0.000872           0.005452      Baxter Intl Inc   
3      -0.064113          -0.011776          -0.052337         Big Lots Inc   
4       0.022305          -0.016412           0.038717      Best Buy Co Inc   

                 industry  direction  
0        Consumer Staples       DOWN  
1             Health Care         UP  
2             Health Care       STAY  
3  Consumer Discretionary       DOWN  
4  Consumer Discretionary         UP  

[5 rows x 26 columns]

As you can see, the table contains daily open and close data for SP500 stocks. The table also contains some features that have been generated for you using [navigation functions](https://cloud.google.com/bigquery/docs/reference/standard-sql/navigation_functions) and [analytic functions](https://cloud.google.com/bigquery/docs/reference/standard-sql/analytic-function-concepts). Let's dig into the schema a bit more. 

In [3]:
%%bigquery 
SELECT
    * EXCEPT(is_generated, generation_expression, is_stored, is_updatable)
FROM
    `cloud-training-prod-bucket.ml4f`.INFORMATION_SCHEMA.COLUMNS
WHERE
    table_name = "percent_change_sp500"

table_catalog table_schema            table_name  \
0   cloud-training-prod-bucket         ml4f  percent_change_sp500   
1   cloud-training-prod-bucket         ml4f  percent_change_sp500   
2   cloud-training-prod-bucket         ml4f  percent_change_sp500   
3   cloud-training-prod-bucket         ml4f  percent_change_sp500   
4   cloud-training-prod-bucket         ml4f  percent_change_sp500   
5   cloud-training-prod-bucket         ml4f  percent_change_sp500   
6   cloud-training-prod-bucket         ml4f  percent_change_sp500   
7   cloud-training-prod-bucket         ml4f  percent_change_sp500   
8   cloud-training-prod-bucket         ml4f  percent_change_sp500   
9   cloud-training-prod-bucket         ml4f  percent_change_sp500   
10  cloud-training-prod-bucket         ml4f  percent_change_sp500   
11  cloud-training-prod-bucket         ml4f  percent_change_sp500   
12  cloud-training-prod-bucket         ml4f  percent_change_sp500   
13  cloud-training-prod-bucket         ml4f  percent_change_sp500   
14  cloud-training-prod-bucket         ml4f  percent_change_sp500   
15  cloud-training-prod-bucket         ml4f  percent_change_sp500   
16  cloud-training-prod-bucket         ml4f  percent_change_sp500   
17  cloud-training-prod-bucket         ml4f  percent_change_sp500   
18  cloud-training-prod-bucket         ml4f  percent_change_sp500   
19  cloud-training-prod-bucket         ml4f  percent_change_sp500   
20  cloud-training-prod-bucket         ml4f  percent_change_sp500   
21  cloud-training-prod-bucket         ml4f  percent_change_sp500   
22  cloud-training-prod-bucket         ml4f  percent_change_sp500   
23  cloud-training-prod-bucket         ml4f  percent_change_sp500   
24  cloud-training-prod-bucket         ml4f  percent_change_sp500   
25  cloud-training-prod-bucket         ml4f  percent_change_sp500   

                    column_name  ordinal_position is_nullable       data_type  \
0                        symbol                 1         YES          STRING   
1                          Date                 2         YES            DATE   
2                          Open                 3         YES         FLOAT64   
3                         Close                 4         YES         FLOAT64   
4                tomorrow_close                 5         YES         FLOAT64   
5            tomo_close_m_close                 6         YES         FLOAT64   
6        close_MIN_prior_5_days                 7         YES         FLOAT64   
7       close_MIN_prior_20_days                 8         YES         FLOAT64   
8      close_MIN_prior_260_days                 9         YES         FLOAT64   
9        close_MAX_prior_5_days                10         YES         FLOAT64   
10      close_MAX_prior_20_days                11         YES         FLOAT64   
11     close_MAX_prior_260_days                12         YES         FLOAT64   
12       close_AVG_prior_5_days                13         YES         FLOAT64   
13      close_AVG_prior_20_days                14         YES         FLOAT64   
14     close_AVG_prior_260_days                15         YES         FLOAT64   
15    close_STDDEV_prior_5_days                16         YES         FLOAT64   
16   close_STDDEV_prior_20_days                17         YES         FLOAT64   
17  close_STDDEV_prior_260_days                18         YES         FLOAT64   
18       close_values_prior_260                19          NO  ARRAY<FLOAT64>   
19               days_on_market                20         YES           INT64   
20                scaled_change                21         YES         FLOAT64   
21            s_p_scaled_change                22         YES         FLOAT64   
22            normalized_change                23         YES         FLOAT64   
23                      company                24         YES          STRING   
24                     industry                25         YES          STRING   
25                    direction                

Most of the features, like `open` and `close` are pretty straightforward. The features generated using analytic functions, such as `close_MIN_prior_5_days` are best described using an example. Let's take the 6 most recent rows of data for IBM and reproduce the `close_MIN_prior_5_days` column. 

In [4]:
%%bigquery
SELECT 
    *
FROM
    `cloud-training-prod-bucket.ml4f.percent_change_sp500`
WHERE
    symbol = 'IBM'
ORDER BY 
    Date DESC
LIMIT 6

symbol        Date    Open   Close  tomorrow_close  tomo_close_m_close  \
0    IBM  2013-02-01  204.65  205.18             NaN                 NaN   
1    IBM  2013-01-31  203.32  203.07          205.18                2.11   
2    IBM  2013-01-30  203.69  203.52          203.07               -0.45   
3    IBM  2013-01-29  204.34  203.90          203.52               -0.38   
4    IBM  2013-01-28  204.85  204.93          203.90               -1.03   
5    IBM  2013-01-25  204.45  204.97          204.93               -0.04   

   close_MIN_prior_5_days  close_MIN_prior_20_days  close_MIN_prior_260_days  \
0                0.989716                 0.937323                  0.879813   
1                1.002216                 0.947063                  0.888955   
2                1.001867                 0.941185                  0.884434   
3                0.961648                 0.930996                  0.878666   
4                0.948958                 0.926316                  0.874250   
5                0.944772                 0.926136                  0.874079   

   close_MAX_prior_5_days  ...  close_STDDEV_prior_20_days  \
0                0.998977  ...                    0.025808   
1                1.009356  ...                    0.025220   
2                1.007125  ...                    0.024643   
3                1.005248  ...                    0.023987   
4                1.000195  ...                    0.021542   
5                0.998780  ...                    0.018510   

   close_STDDEV_prior_260_days  \
0                     0.031267   
1                     0.031925   
2                     0.032221   
3                     0.032552   
4                     0.032677   
5                     0.032874   

                              close_values_prior_260  days_on_market  \
0  [180.52, 188.52, 189.98, 191.93, 191.73, 190.9...           12860   
1  [181.07, 180.52, 188.52, 189.98, 191.93, 191.7...           12859   
2  [180.0, 181.07, 180.52, 188.52, 189.98, 191.93...           12858   
3  [179.16, 180.0, 181.07, 180.52, 188.52, 189.98...           12857   
4  [180.55, 179.16, 180.0, 181.07, 180.52, 188.52...           12856   
5  [182.32, 180.55, 179.16, 180.0, 181.07, 180.52...           12855   

   scaled_change  s_p_scaled_change  normalized_change  \
0            NaN          -0.011539                NaN   
1       0.010391           0.010053           0.000338   
2      -0.002211          -0.002563           0.000352   
3      -0.001864          -0.003900           0.002036   
4      -0.005026           0.005106          -0.010132   
5      -0.000195          -0.001850           0.001655   

                       company                industry  direction  
0  Intl Business Machines Corp  Information Technology       STAY  
1  Intl Business Machines Corp  Information Technology       STAY  
2  Intl Business Machines Corp  Information Technology       STAY  
3  Intl Business Machines Corp  Information Technology       STAY  
4  Intl Business Machines Corp  Information Technology       DOWN  
5  Intl Business Machines Corp  Information Technology       STAY  

[6 rows x 26 columns]

For `Date = 2013-02-01` how did we arrive at `close_MIN_prior_5_days = 0.989716`? The minimum close over the past five days was `203.07`. This is normalized by the current day's close of `205.18` to get `close_MIN_prior_5_days = 203.07 / 205.18 = 0.989716`. The other features utilizing analytic functions were generated in a similar way. Here are explanations for some of the other features:

* __scaled_change__: `tomo_close_m_close / close`
* __s_p_scaled_change__: This value is calculated the same way as `scaled_change` but for the S&P 500 index. 
* __normalized_change__: `scaled_change - s_p_scaled_change` The normalization using the S&P index fund helps ensure that the future price of a stock is not due to larger market effects. Normalization helps us isolate the factors contributing to the performance of a stock_market.
* __direction__: This is the target variable we're trying to predict. The logic for this variable is as follows: 

    ```sql
    CASE 
        WHEN normalized_change < -0.01 THEN 'DOWN'
        WHEN normalized_change > 0.01 THEN 'UP'
        ELSE 'STAY'
    END AS direction
    ```

## Create classification model for `direction`

In this example, your job is to create a classification model to predict the `direction` of each stock. Be creative! You can do this in any number of ways. For example, you can use BigQuery, Scikit-Learn, or AutoML. Feel free to add additional features, or use time series models.   



### Establish a Simple Benchmark

One way to assess the performance of a model is to compare it to a simple benchmark. We can do this by seeing what kind of accuracy we would get using the naive strategy of just predicting the majority class. Across the entire dataset, the majority class is 'STAY'. Using the following query we can see how this naive strategy would perform.

In [5]:
%%bigquery
WITH subset as (
    SELECT 
        Direction
    FROM
        `cloud-training-prod-bucket.ml4f.percent_change_sp500`
    WHERE
        tomorrow_close IS NOT NULL
)
SELECT 
    Direction,
    100.0 * COUNT(*) / (SELECT COUNT(*) FROM subset) as percentage
FROM
    subset
GROUP BY
    Direction

Direction  percentage
0      STAY   53.766049
1        UP   23.240681
2      DOWN   22.993271

So, the naive strategy of just guessing the majority class would have accuracy of around 54% across the entire dataset. See if you can improve on this. 

### Train Your Own Model

In [6]:
%%bigquery 
CREATE OR REPLACE MODEL `ai4f.directionDNNClassifier`
OPTIONS
  (model_type='DNN_CLASSIFIER', input_label_cols=['direction']) AS
SELECT
  symbol,
  close_MIN_prior_5_days,
  close_MIN_prior_20_days,
  close_MAX_prior_5_days,
  close_MAX_prior_20_days,
  direction
FROM
  `cloud-training-prod-bucket.ml4f.percent_change_sp500`
WHERE
  tomorrow_close IS NOT NULL
  AND EXTRACT(year FROM Date) = 2013

Empty DataFrame
Columns: []
Index: []

In [7]:
%%bigquery dfEval

SELECT * FROM ML.EVALUATE(MODEL ai4f.directionDNNClassifier,
(
SELECT
  symbol,
  close_MIN_prior_5_days,
  close_MIN_prior_20_days,
  close_MAX_prior_5_days,
  close_MAX_prior_20_days,
  direction
FROM
  `cloud-training-prod-bucket.ml4f.percent_change_sp500`
WHERE
  tomorrow_close IS NOT NULL
  AND EXTRACT(year FROM Date) = 2012
))

In [9]:
dfEval

precision    recall  accuracy  f1_score  log_loss   roc_auc
0   0.310904  0.299175     0.333  0.270956  1.153463  0.502827

In [10]:
%%bigquery dfPredict

SELECT * FROM ML.PREDICT(MODEL ai4f.directionDNNClassifier,
(
SELECT
  Date,
  symbol,
  close_MIN_prior_5_days,
  close_MIN_prior_20_days,
  close_MAX_prior_5_days,
  close_MAX_prior_20_days,
  direction AS ActualDirection
FROM
  `cloud-training-prod-bucket.ml4f.percent_change_sp500`
WHERE
  tomorrow_close IS NOT NULL
  AND EXTRACT(year FROM Date) = 2011
  AND EXTRACT(month FROM Date) = 12
  AND symbol = 'IBM'
))

In [11]:
dfPredict

predicted_direction                          predicted_direction_probs  \
0                 STAY  [{'label': 'UP', 'prob': 0.2722116708755493}, ...   
1                 DOWN  [{'label': 'UP', 'prob': 0.24072584509849548},...   
2                 STAY  [{'label': 'UP', 'prob': 0.30264806747436523},...   
3                 DOWN  [{'label': 'UP', 'prob': 0.23078176379203796},...   
4                 STAY  [{'label': 'UP', 'prob': 0.29102766513824463},...   
5                 STAY  [{'label': 'UP', 'prob': 0.3012154698371887}, ...   
6                 STAY  [{'label': 'UP', 'prob': 0.34357231855392456},...   
7                 STAY  [{'label': 'UP', 'prob': 0.3395346403121948}, ...   
8                 STAY  [{'label': 'UP', 'prob': 0.33943337202072144},...   
9                 DOWN  [{'label': 'UP', 'prob': 0.25525403022766113},...   
10                STAY  [{'label': 'UP', 'prob': 0.31488296389579773},...   
11                STAY  [{'label': 'UP', 'prob': 0.33380329608917236},...   
12                STAY  [{'label': 'UP', 'prob': 0.30099979043006897},...   
13                STAY  [{'label': 'UP', 'prob': 0.3055739402770996}, ...   
14                DOWN  [{'label': 'UP', 'prob': 0.20275470614433289},...   
15                DOWN  [{'label': 'UP', 'prob': 0.2769043445587158}, ...   
16                STAY  [{'label': 'UP', 'prob': 0.3300633430480957}, ...   
17                STAY  [{'label': 'UP', 'prob': 0.27261054515838623},...   
18                DOWN  [{'label': 'UP', 'prob': 0.24231092631816864},...   
19                STAY  [{'label': 'UP', 'prob': 0.310487300157547}, {...   
20                STAY  [{'label': 'UP', 'prob': 0.3136797547340393}, ...   

          Date symbol  close_MIN_prior_5_days  close_MIN_prior_20_days  \
0   2011-12-22    IBM                0.996869                 0.972643   
1   2011-12-29    IBM                0.974702                 0.974702   
2   2011-12-15    IBM                1.006614                 0.944421   
3   2011-12-28    IBM                0.986304                 0.983423   
4   2011-12-20    IBM                0.976768                 0.945631   
5   2011-12-19    IBM                1.003718                 0.968123   
6   2011-12-07    IBM                0.968822                 0.912445   
7   2011-12-06    IBM                0.937804                 0.917695   
8   2011-12-09    IBM                0.974815                 0.910053   
9   2011-12-21    IBM                1.007825                 0.975698   
10  2011-12-02    IBM                0.933565                 0.933565   
11  2011-12-05    IBM                0.948124                 0.927793   
12  2011-12-13    IBM                1.002250                 0.926288   
13  2011-12-23    IBM                0.982246                 0.958376   
14  2011-12-30    IBM                0.989993                 0.986894   
15  2011-12-14    IBM                1.012876                 0.938215   
16  2011-12-01    IBM                0.934600                 0.934600   
17  2011-12-16    IBM                1.021300                 0.964537   
18  2011-12-27    IBM                0.981184                 0.978318   
19  2011-12-08    IBM                0.988882                 0.924209   
20  2011-12-12    IBM                0.993027                 0.921324   

    close_MAX_prior_5_days  close_MAX_prior_20_days ActualDirection  
0                 1.029884                 1.068776            STAY  
1                 0.993393                 1.045010            STAY  
2                 1.037764                 1.037764            DOWN  
3                 1.017664                 1.057449            STAY  
4                 1.020882                 1.039094            DOWN  
5                 1.050796                 1.063809            STAY  
6                 0.994280                 0.994280            STAY  
7                 0.989116                 0.989116            STAY  
8                 0.997379                 0.9973